### PR TITLE
Add IPC3 Message Injector

### PR DESCRIPTION
**Pull request by cujomalainey**
Add the ability to fuzz the kernel IPC interface via a sysfs path

IPC4 support will be added when IPC4 begins consuming its data in its RX handler

### CODE REVIEW

File: sound/soc/sof/ipc3.c
Comment (plbossart) [inline sound/soc/sof/ipc3.c:1051]: I don't understand why we need all this.

The client injector should do what you want already. What is missing in its existing implementation?

Can you clarify the purpose of this patch?

Comment (cujomalainey) [reply sound/soc/sof/ipc3.c:1051]: The client injector transmits to the firmware, this splits the injector into another path to inject into the kernel so messages injected come in as if they were FROM the fw.

Comment (cujomalainey) [reply sound/soc/sof/ipc3.c:1051]: Also we have a requirement on our side to fuzz the kernel to harden it against malicious firmware, which is why we are adding this.

Comment (plbossart) [reply sound/soc/sof/ipc3.c:1051]: oh so you use the client injector to send stuff to the firmware but override the replies from the firmware with your side path?

Comment (ranj063) [reply sound/soc/sof/ipc3.c:1051]: @plbossart I think there's no TX here at all. I think they're trying to fuzz only the rx from the DSP part without any IPCs sent to the DSP

**Comment by ranj063** [comment]
@cujomalainey could you please take care of the checkpatch warning?

File: sound/soc/sof/ipc3.c
Comment (cujomalainey) [reply sound/soc/sof/ipc3.c:1051]: @ranj063 is correct, this is fuzzing a DSP originated message, replies are not in scope currently.

Comment (plbossart) [reply sound/soc/sof/ipc3.c:1051]: Not sure I understand the scope of the testing then. If there is nothing really consuming the information, what does this accomplish? Is this just to test the DSP notification and DSP-initiated alerts? Isn't that a very very small part of what the IPC protocol is about.

I would have sent a fake message and inserted a fake reply to cause all sorts of trouble.

Comment (cujomalainey) [reply sound/soc/sof/ipc3.c:1051]: @plbossart this is a requirement by our security team, and yes replies would cover more ground, which we will look into, but for right now we are starting here.

It may be a small part of the code, but there are still [consumers](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/ipc3.c#L1029) of the data.

Comment (ranj063) [reply sound/soc/sof/ipc3.c:1051]: @plbossart I think the scope of fuzzing can include fake messages injection in both directions but done independently. With this PR, the injector client can handle both.

Comment (cujomalainey) [reply sound/soc/sof/ipc3.c:1051]: also we rarely use replies

**Comment by ranj063** [comment]
> @cujomalainey could you please take care of the checkpatch warning?

@cujomalainey you've got a different set of checkpatch warnings now

**Comment by cujomalainey** [comment]
> > @cujomalainey could you please take care of the checkpatch warning?
> 
> @cujomalainey you've got a different set of checkpatch warnings now

Gah, its been so long since I have sent a patch, is there a way I can run this out of tree? I tried it directly with scripts/checkpatch.pl but got different warnings than what CI is giving.

**Comment by ranj063** [comment]
> > > @cujomalainey could you please take care of the checkpatch warning?
> > 
> > 
> > @cujomalainey you've got a different set of checkpatch warnings now
> 
> Gah, its been so long since I have sent a patch, is there a way I can run this out of tree? I tried it directly with scripts/checkpatch.pl but got different warnings than what CI is giving.

This is what I use in my git hooks

``` 
:~/asoc/.git/hooks$ cat pre-commit
#!/bin/bash
# SPDX-License-Identifier: BSD-3-Clause
# Copyright(c) 2018 Intel Corporation. All rights reserved.

exec git diff --cached | scripts/checkpatch.pl --no-tree --codespell --no-signoff -q -

:~/asoc/.git/hooks$ cat post-commit
#!/bin/sh
# SPDX-License-Identifier: BSD-3-Clause
# Copyright(c) 2018 Intel Corporation. All rights reserved.

set -e
# TODO: reduce duplication with scripts/sof-pre-commit-hook.sh
# and with .github/workflows/ with either some .conf file
# or some wrapper script
exec git show --format=email HEAD |
 ./scripts/checkpatch.pl --no-tree --strict --codespell --ignore C99_COMMENT_TOLERANCE
```

**Comment by cujomalainey** [comment]
Thanks this is likely because I have this checkedout within the chromeos tree

**Review by ranj063 — DISMISSED** [review]

File: sound/soc/sof/sof-client-ipc-msg-injector.c
Comment (ujfalusi) [inline sound/soc/sof/sof-client-ipc-msg-injector.c]: SOF client drivers must never use 'sdev'

File: sound/soc/sof/sof-client-ipc-msg-injector.c
Comment (ujfalusi) [inline sound/soc/sof/sof-client-ipc-msg-injector.c]: subject: "ASoC: SOF: ipc-msg-injector: Add support for injecting message for the kernel"

File: sound/soc/sof/sof-client-ipc-msg-injector.c
Comment (ujfalusi) [inline sound/soc/sof/sof-client-ipc-msg-injector.c]: SOF client drivers must never use SOF core functionality.

File: sound/soc/sof/sof-client-ipc-msg-injector.c
Comment (ujfalusi) [inline sound/soc/sof/sof-client-ipc-msg-injector.c]: Given that this feature basically adds a 'fake' DSP which sends messages to the kernel, I think it would be better to have a new Kconfig option with a new client driver and have the needed interfaces (sof-clientized!) be #idfef wrapped.

**Review by ujfalusi — CHANGES_REQUESTED** [review]

File: sound/soc/sof/sof-client-ipc-msg-injector.c
Comment (cujomalainey) [reply sound/soc/sof/sof-client-ipc-msg-injector.c]: `sdev` is used indirectly via `sof_client_ipc_tx_message` would you rather I shove the callback logic into `sof-client.c`?

File: sound/soc/sof/sof-client-ipc-msg-injector.c
Comment (cujomalainey) [reply sound/soc/sof/sof-client-ipc-msg-injector.c]: That would require taking any common code and moving to a third file, what is the motive behind splitting this from the existing injector?

**Comment by cujomalainey** [comment]
@ujfalusi ping on questions

File: sound/soc/sof/sof-client-ipc-msg-injector.c
Comment (ujfalusi) [reply sound/soc/sof/sof-client-ipc-msg-injector.c]: The `sof_client_ipc_tx_message()` is within snd_sof.ko it can use sdev, the client drivers are _not_ within the core SOF modules and they must not.

File: sound/soc/sof/sof-client-ipc-msg-injector.c
Comment (ujfalusi) [reply sound/soc/sof/sof-client-ipc-msg-injector.c]: I don't really see common code here apart from the mandatory sof-client/auxiliary driver stuff and extension of a struct used for the message injection.

This is really a feature which is completely independent and has no relation to the message injection client _to_ the firmware.

Comment (ujfalusi) [reply sound/soc/sof/sof-client-ipc-msg-injector.c]: In the 'fake' DSP case you only need the write debugfs file, I don't know what you would do with the read in case you torture the Kernel...

**Review by ujfalusi — CHANGES_REQUESTED** [review]
@cujomalainey, I don't think it is a good idea to extend the current (firmware) message injector with a functionality completely different and unrelated.

It is going to be cleaner and neater to have a new client driver to implement a fake DSP which sends notifications to the kernel.

File: sound/soc/sof/sof-client-ipc-kernel-injector.c
Comment (ranj063) [inline sound/soc/sof/sof-client-ipc-kernel-injector.c]: do you need this include? it doesnt look like IPC4 is supported by this client just yet

Comment (cujomalainey) [reply sound/soc/sof/sof-client-ipc-kernel-injector.c]: Ah thanks for catching my copy paste error

Comment (cujomalainey) [reply sound/soc/sof/sof-client-ipc-kernel-injector.c]: Actually this is needed, since the injector itself supports IPC4, it is the sof-client that does not, so I think I will leave it for the alloc modification

**Review by ranj063 — DISMISSED** [review]

**Comment by cujomalainey** [comment]
Pushed a small fix to clean some other headers and add a error check

**Review by ranj063 — DISMISSED** [review]

File: sound/soc/sof/ipc3-priv.h
Comment (ujfalusi) [inline sound/soc/sof/ipc3-priv.h:32]: This comment is not quite accurate, the `sof_ipc3_do_rx_work()` is used by `sof_ipc3_rx_msg()` at this point only and it is the actual rx handler.

File: sound/soc/sof/sof-client-ipc-kernel-injector.c
Comment (ujfalusi) [inline sound/soc/sof/sof-client-ipc-kernel-injector.c]: The injector should not be loaded in the first place when IPC4 is used as it is IPC3 only at the moment.

File: sound/soc/sof/sof-client.c
Comment (ujfalusi) [inline sound/soc/sof/sof-client.c:132]:
```c
if (sdev->pdata->ipc_type != SOF_IPC)
	return 0;
```
To block the fake DSP to be instantiated when IPC4 is used?

**Review by ujfalusi — COMMENTED** [review]
@cujomalainey, this looks much better!
Can you check and if you see fit, address my two comments?

File: sound/soc/sof/sof-client.c
Comment (cujomalainey) [reply sound/soc/sof/sof-client.c:132]: Yea, IPC4 has no RX handlers last I looked, I will add it once there is something to actually parse the data

Comment (ujfalusi) [reply sound/soc/sof/sof-client.c:132]: can you update the PR to not create the aux device in case of IPC4? We can enable the fake DSP when it is supporting IPC4.

**Comment by plbossart** [comment]
@cujomalainey can you rebase and address @ujfalusi's comment if you have a bit of time? I'd like to push this PR in the current kernel cycle, thanks!

**Comment by cujomalainey** [comment]
@plbossart apologies, review season plus was remote for the holidays, will pick this up next week

**Comment by cujomalainey** [comment]
device is now blocked on IPC4

**Comment by cujomalainey** [comment]
ping to @ujfalusi

File: sound/soc/sof/sof-client.c
Comment (ujfalusi) [reply sound/soc/sof/sof-client.c:132]: I still think that the aux device should not be created in the first place
```c
if (sdev->pdata->ipc_type != SOF_IPC)
		return 0;
```
and the injector driver then can be free form IPC version check for now as it is only going to probe with a known IPC version.

File: sound/soc/sof/ipc3.c
Comment (ujfalusi) [inline sound/soc/sof/ipc3.c:1051]: Nitpick: empty line between the closing curly-bracket and the function call?

File: sound/soc/sof/sof-client-ipc-kernel-injector.c
Comment (cujomalainey) [inline sound/soc/sof/sof-client-ipc-kernel-injector.c]: @ujfalusi yes the device creation is aborted here

Comment (ujfalusi) [reply sound/soc/sof/sof-client-ipc-kernel-injector.c]: no, this is the driver probe, the device is created in sof-client.c at line 133 and there you should not create the device in the first place.

Comment (cujomalainey) [reply sound/soc/sof/sof-client-ipc-kernel-injector.c]: Sorry this took so long, patch is fixed

File: sound/soc/sof/Kconfig
Comment (ujfalusi) [inline sound/soc/sof/Kconfig:249]: what does this means in the commit message?
`Skip IPC4 as rx messages are not currently parsed outside of their registers.`

IPC4 handles SOF_IPC4_NOTIFY_FW_READY, SOF_IPC4_NOTIFY_RESOURCE_EVENT and SOF_IPC4_NOTIFY_LOG_BUFFER_STATUS, but it is true that only FW_READY and LOG_BUFFER_STATUS have real users.

File: sound/soc/sof/sof-client-ipc-kernel-injector.c
Comment (ujfalusi) [inline sound/soc/sof/sof-client-ipc-kernel-injector.c]: 2023?

File: sound/soc/sof/sof-client-ipc-kernel-injector.c
Comment (ujfalusi) [reply sound/soc/sof/sof-client-ipc-kernel-injector.c]: Hrm, I think this fake DSP is _not_ supporting IPC4, there is a huge difference between how IPC3 and IPC4 messages are stored in the kernel and you cannot just copy a blob from user space to a buffer and pass that as `struct sof_ipc4_msg`, the payload (on top of the header) is behind of a pointer (data_ptr) and there is a data_size as well.

It is not right to claim that this will work with IPC4 since it is not and will crash the kernel.

There is no need to include the ipc4 header or have a proper implementation for IPC4 in the driver.

**Comment by plbossart** [comment]
@cujomalainey @ujfalusi what's the direction for this PR? It doesn't seem like we have agreement here, do we?

**Comment by cujomalainey** [comment]
It's blocked on me and I have been blocked internally on other work, will get back to it soon

File: sound/soc/sof/Kconfig
Comment (cujomalainey) [reply sound/soc/sof/Kconfig:249]: at the time of writing all IPC4 messages only read status bit and did not parse the payload, if that is not true anymore then i will update

File: sound/soc/sof/sof-client-ipc-kernel-injector.c
Comment (cujomalainey) [reply sound/soc/sof/sof-client-ipc-kernel-injector.c]: @ujfalusi that is the point. The kernel should not trust the data and should reject malformed payloads. 

And yes there will be a IPC4 implementation. Regardless of semantics, we have security requirements to fuzz this interface.

File: sound/soc/sof/Kconfig
Comment (cujomalainey) [reply sound/soc/sof/Kconfig:249]: Nope the statement is still true, all these message types only read the registers

```
static void sof_ipc4_rx_msg(struct snd_sof_dev *sdev)
{
...
	if (data_size) {
		ipc4_msg->data_ptr = kmalloc(data_size, GFP_KERNEL);
		if (!ipc4_msg->data_ptr)
			return;

		ipc4_msg->data_size = data_size;
		snd_sof_ipc_msg_data(sdev, NULL, ipc4_msg->data_ptr, ipc4_msg->data_size);
	}

	sof_ipc4_log_header(sdev->dev, "ipc rx done ", ipc4_msg, true);

	if (data_size) {
		kfree(ipc4_msg->data_ptr);
		ipc4_msg->data_ptr = NULL;
		ipc4_msg->data_size = 0;
	} 
```

The data is read and then freed right after without any call to a handler

File: sound/soc/sof/sof-client.c
Comment (ujfalusi) [inline sound/soc/sof/sof-client.c]: nitpick: the reason that I suggested:
```c
if (sdev->pdata->ipc_type != SOF_IPC)
	return 0;

return sof_client_dev_register(sdev, "kernel_injector", 0, NULL, 0);
```
is that this way we can end up with shorter lines and aligned with how `sof_register_ipc_flood_test()` has been implemented.

**Review by ujfalusi — DISMISSED** [review]
@cujomalainey, there is only one minor nitpick left in my nitpick bag ;)

**Comment by cujomalainey** [comment]
@plbossart @ranj063 I think we are ready for final review, apologies on the long timeline

**Review by ranj063 — APPROVED** [review]
LGTM

**Review by ranj063 — DISMISSED** [review]
LGTM

File: sound/soc/sof/sof-client-ipc-kernel-injector.c
Comment (plbossart) [inline sound/soc/sof/sof-client-ipc-kernel-injector.c:9]: commit message: Add sysfs path

You really meant 'Add debugfs path'

If you want sysfs, then we need to document the ABI and run it past GregKH, different process.

File: sound/soc/sof/sof-client-ipc-kernel-injector.c
Comment (plbossart) [inline sound/soc/sof/sof-client-ipc-kernel-injector.c:122]: don't we need a pm_runtime_set_active(dev); before the pm_runtime_enable()?

**Review by ujfalusi — DISMISSED** [review]
@cujomalainey, my nitpick bag is now empty, thank you for being persistent!

**Review by plbossart — CHANGES_REQUESTED** [review]
sorry @cujomalainey somehow GitHub did not post my previous comments, see minor nit-picks below.

File: sound/soc/sof/sof-client-ipc-kernel-injector.c
Comment (cujomalainey) [reply sound/soc/sof/sof-client-ipc-kernel-injector.c:122]: no gonna lie, I am the among the biggest of noobs when it comes to PM so that is probably true.

File: sound/soc/sof/sof-client-ipc-kernel-injector.c
Comment (cujomalainey) [reply sound/soc/sof/sof-client-ipc-kernel-injector.c:9]: Ack will fix, thanks for catching

File: sound/soc/sof/sof-client-ipc-kernel-injector.c
Comment (plbossart) [inline sound/soc/sof/sof-client-ipc-kernel-injector.c]: you also want a pm_runtime_resume_and_get() before sending an IPC and a pm_runtime_mark_last_busy() + pm_runtime_put_autosuspend() after the IPC?

See e.g. the SoundWire example here: https://github.com/thesofproject/linux/blob/topic/sof-dev/drivers/soundwire/debugfs.c#L59 

https://github.com/thesofproject/linux/blob/topic/sof-dev/drivers/soundwire/debugfs.c#L123

Comment (cujomalainey) [reply sound/soc/sof/sof-client-ipc-kernel-injector.c]: actually do we need PM? This fs is not interacting with the HW in anything way.

Comment (plbossart) [reply sound/soc/sof/sof-client-ipc-kernel-injector.c]: if you application that writes into debugfs lets the DSP suspend, you'll never get any IPC sent. Somehow you need to resume the DSP then send the IPC.

The client to inject the message is a child of the PCI device, so if you pm_runtime_resume it the PCI is also resumed. Conversely the PCI device will only suspend if you application stops sending messages for more than 3s.

Comment (cujomalainey) [reply sound/soc/sof/sof-client-ipc-kernel-injector.c]: The IPC isn't going into the DSP though. Its going back into the kernel. The IPC is being sent falsely from the DSP to provide an interface to fuzz the kernel

Comment (plbossart) [reply sound/soc/sof/sof-client-ipc-kernel-injector.c]: what happens if the driver wants to dump a state or register which isn't powered?

Notifications from the DSP can only come when the DSP is powered, I would err on the side of avoiding to add more variables...

Comment (cujomalainey) [reply sound/soc/sof/sof-client-ipc-kernel-injector.c]: Good point, we should force the DSP awake if we are going to pretend to be it as we won't know what we will trigger

File: sound/soc/sof/sof-client-ipc-kernel-injector.c
Comment (plbossart) [inline sound/soc/sof/sof-client-ipc-kernel-injector.c]: you need to release the reference taken line 60

**Review by plbossart — CHANGES_REQUESTED** [review]

File: sound/soc/sof/sof-client-ipc-kernel-injector.c
Comment (cujomalainey) [reply sound/soc/sof/sof-client-ipc-kernel-injector.c]: isnt' that what `pm_runtime_put_autosuspend` does?

Comment (cujomalainey) [reply sound/soc/sof/sof-client-ipc-kernel-injector.c]: nvm I see your point

Comment (cujomalainey) [reply sound/soc/sof/sof-client-ipc-kernel-injector.c]: removed the error path since it doesn't matter to the fuzzer

**Review by plbossart — APPROVED** [review]
LGTM but @ujfalusi needs to review as well.

**Review by ujfalusi — APPROVED** [review]

**Pull request merged**
